### PR TITLE
chore(ci): add jobs for testing across mongod versions MONGOSH-312

### PR DESCRIPTION
**Pull request by addaleax**
This is basically a revamp the evergreen config:

- Rename CI tasks to be more descriptive
- Add CI tasks to verify that the Linux artifacts actually run on
  their target platforms (we previously had jobs in CI that were
  *named* to imply that they were using RHEL 8.0 and Debian 10
  but did not actually run those OSes in any way, only built
  packages targeting those)
- Add test_m??x tasks where ?? represents the target server version
  for testing
- Update the mongod download script to account for those different
  server versions across the supported OSes
- Move darwin testing off the codesign machine (because there is
  only one host for that) – compiling still takes place there
- Update the build config script to account for the fact that
  the `revision` expansion only represents the *base* revision for
  patches, and use a different format for patches so that S3
  artifacts do not accidentally get overridden with new artifacts
  from patches targeting the same base revision
- Actually wait for all other tasks to succeed before continuing
  with a release

### CODE REVIEW

File: .evergreen.yml [hunk 1]
```diff
@@ -218,115 +226,173 @@ tasks:
...
-      - name: test
+      - name: test_m40x
+      - name: test_m42x
+      - name: test_m44x
       - name: test_vscode
-      - name: release_linux
+      - name: compile_and_upload_artifact_linux
   - name: rhel
-    display_name: "RHEL 8.0 (in Ubuntu docker)"
+    display_name: "Ubuntu 18.04 (rpm target)"
```
Comment (rose-m) [inline .evergreen.yml:366]: Stupid question: what are `rhel` and `debian`? They are both running on `ubuntu1804-test` and thus the same as `linux`? 🤔 Can we throw them out?

**Review by rose-m — APPROVED** [review]
Testing all versions 👏 😄 Just got a minor question regarding `rhel` vs `rhel_real` that's got me like 🤔

File: .evergreen.yml [hunk 2]
```diff
@@ -151,36 +171,24 @@ functions:
             -e BUILD_VARIANT \
             --rm -v $PWD:/tmp/build ubuntu18.04-xvfb \
             -c 'cd /tmp/build && ./testing/test-vscode.sh'
-  release_macos:
+  compile_and_upload_artifact:
```
Comment (mcasimir) [inline .evergreen.yml:174]: Nice! Super clear

File: .evergreen.yml [hunk 3]
```diff
@@ -218,115 +226,173 @@ tasks:
...
+      - name: compile_and_upload_artifact_linux
+        variant: linux
+    commands:
+      - func: checkout
+      - func: install
+      - func: test_linux_artifact
   - name: release_publish
     depends_on:
-      - name: release_win_ps
+      - name: compile_and_upload_artifact_ps
```
Comment (mcasimir) [inline .evergreen.yml:323]: Probably at this point we can rename `ps` to `win` and `powershell` to `windows`?
I think the only reason we had `ps` was cause at some point we were transitioning from `cygwin` which was the default windows shell in evergreen

File: config/build.conf.js [hunk 4]
```diff
@@ -1,3 +1,5 @@
+'use strict';
```
Comment (mcasimir) [inline config/build.conf.js:1]: just curious, why is this needed here?

File: config/build.conf.js [hunk 5]
```diff
@@ -41,6 +43,13 @@ const ANALYTICS_CONFIG = path.join(MONGOSH, 'lib', 'analytics-config.js');
  */
 const APPLE_NOTARIZATION_BUNDLE_ID = 'com.mongodb.mongosh';
 
+/**
+ * The SHA for the current git HEAD.
+ */
+const REVISION = process.env.IS_PATCH ?
+  `pr-${process.env.GITHUB_PR_NUMBER}-${process.env.REVISION_ORDER_ID}` :
```
Comment (mcasimir) [inline config/build.conf.js:50]: How is this used? This will be probably super-useful for releases too.

**Review by mcasimir — APPROVED** [review]
👍

File: .evergreen.yml (hunk 1 above)
Comment (addaleax) [reply .evergreen.yml:366]: Ah, right :slightly_smiling_face: So … our `@mongosh/build` package makes a distinction on how to behave based on the build variant name, and, in particular, it packages as a tarball/deb/rpm based on it. The variable gets picked up in `config/build.conf.js` from an environment variable, and passed around throughout the build package.

That’s why the names here, `rhel` and `debian`, are not freely changeable:

https://github.com/mongodb-js/mongosh/blob/dde7b239f0cef50d217a4bf1766a89c8890a2e9d/packages/build/src/build-variant.ts#L7-L13

This is obviously not an ideal situation, no. It means that we’re currently building the Linux binary three separate times, only differing in the final packaging step, which wastes lots of CPU time in CI :slightly_smiling_face: But I also wanted to get the specific task in question done, and this doesn’t make anything worse in this respect as far as I can tell.

Ultimately, it would be nice imo if:

- We didn’t make choices on how to behave purely based on the build variant name, and
- We only built the Linux binary exactly once.

File: config/build.conf.js (hunk 4 above)
Comment (addaleax) [reply config/build.conf.js:1]: @mcasimir It’s not needed for anything in particular, it’s just good practice to always write JS code in strict mode (TS should already always be strict mode for us)

File: config/build.conf.js (hunk 5 above)
Comment (addaleax) [reply config/build.conf.js:50]: Err, it might be easier to show this once CI has run :) It’s used for uploading artifacts to S3; e.g. the one from this PR would now become available at 

https://s3.amazonaws.com/mciuploads/mongosh/pr-433-anna.henningsen_158/mongosh-0.5.2-linux.tgz

instead of 

https://s3.amazonaws.com/mciuploads/mongosh/dde7b239f0cef50d217a4bf1766a89c8890a2e9d/mongosh-0.5.2-linux.tgz

(where dde7b239f0cef50d217a4bf1766a89c8890a2e9d is the *base* revision, not a commit from this PR). That matters because we would otherwise overwrite the already-existing tarball for dde7b239f0cef50d217a4bf1766a89c8890a2e9d – we’re lucky we didn’t have problems with this, but in the worst case, I think this could previously have potentially made builds from a PR override a release build :grimacing:

File: .evergreen.yml (hunk 1 above)
Comment (rose-m) [reply .evergreen.yml:366]: Thank you very much! Didn't expect these names to be relevant outside the config, but that's good to know 👍

File: .evergreen.yml (hunk 3 above)
Comment (addaleax) [reply .evergreen.yml]: Sure, done :+1:
